### ISSUE TITLE
Improve error message when invariants don't match (e.g., due to parameter name mismatches)

### ISSUE THREAD

**Issue by bersbersbers** [microsoft/pyright#5822]
**Describe the bug**
I have been playing around for a while until I found the issue with the code below. The reason is a simply mismatch of parameter names, but the error message does not say this. The problem may be obvious in this simple example, but in the case of generic classes and protocols spread out over multiple files, this gets really hard to track down.

Error message:
```
C:\Code\project\bug.py
  C:\Code\project\bug.py:18:21 - error: Argument of type "list[Class]" cannot be assigned to parameter "instances" of type "list[SetValue]" in function "set_many_values"
    "list[Class]" is incompatible with "list[SetValue]"
      Type parameter "_T@list" is invariant, but "Class" is not the same as "SetValue" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 informations
```

In that context, I must say I *love* `pyright`'s hierarchical error messages, and it seems like the just the depth would need to be increased. In fact, the final line in my code example shows what a good error message can look like:

```
[...]
C:\Code\project\bug.py:20:17 - error: Expression of type "Class" cannot be assigned to declared type "SetValue"
    "Class" is incompatible with protocol "SetValue"
      "set_value" is an incompatible type
        Type "(val: int) -> None" cannot be assigned to type "(value: int) -> None"
          Parameter name mismatch: "value" versus "val" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

**Code or Screenshots**
```python
from typing import Protocol

class Class:
    def set_value(self, val: int) -> None:  # note "val"
        ...

class SetValue(Protocol):
    def set_value(self, value: int) -> None:  # note "value" != "val"
        ...

def set_many_values(
    instances: list[SetValue], values: list[int]
) -> None:
    for instance, value in zip(instances, values):
        instance.set_value(value)

def set_shape(instances: list[Class], values: list[int]) -> None:
    set_many_values(instances, values)

# var: SetValue = Class()  # this error message is much more meaningful
```

pyright 1.1.324 via command line

**Comment by erictraut** [comment]
The error in your sample code is not caused by a simple mismatch of parameter names. Even if you change the parameter name from `value` to `val` in the protocol definition, it is still a type violation. As indicated in the error message, the type parameter `_T` for `list` is invariant, which means that the type arguments must match exactly. While `Class` is a subtype of `SetValue` (if you fix the parameter name issue), it is not the _same_ type. I'm not sure how to make the error message any clearer for the invariant case.

If the type variable were covariant or contravariant, I think there would be a better argument for additional information. For example, if you modify your sample above so the parameter `instances` is annotated with the type `Sequence[Class]`, the error message currently says `Type parameter "_T_co@Sequence" is covariant, but "Class" is not a subtype of "SetValue"`.

**Comment by bersbersbers** [comment]
Thanks for the analysis! Yes, my actual code does use `Sequence` - I just incorrectly "simplified" it to `list` when creating my code example to save one import because I was not aware of the difference in variance between `list` and `Sequence`, and missed the change in error message from `is not a subtype of` to `is not the same as`.

The fix you implemented is exactly what I had in mind. Is there any particular reason the line
>           Parameter name mismatch: "value" versus "val" (reportGeneralTypeIssues)
appears in https://github.com/microsoft/pyright/issues/5822#issue-1868241454 but not in https://github.com/microsoft/pyright/issues/5822#issuecomment-1694505308, or is that simply a copy-and-paste error? `addAddendum` in 0240f557643fb44f9af2e466a3f9137219cdf5c3 does not look like it distinguishes between both cases, so I'd expect both versions to report the parameter name mismatch.

**Comment by erictraut** [comment]
Pyright limits error messages to 5 levels of hierarchy and 8 lines total. There must be some limit placed on them, and I think this is a reasonable limit. It's already a larger limit than used in the TypeScript compiler, which also uses hierarchical errors like this. There will always be some error messages that are clipped, but that will be true regardless of which limit is adopted.

**Comment by bersbersbers** [comment]
Makes sense! Maybe `--verbose` could show the full error message - it's the first thing I tried when I got the error message that wasn't clear to me.

**Comment by erictraut** [comment]
This is included in pyright 1.1.325, which I just published. It will also be included in an upcoming pylance release.
